backend/python/tapa/floorplan_dse.py

```diff
@@ -0,0 +1,201 @@
+    '--pack-xo',
+    '--max-slr-width-limit', str(max_slr_limit),
+    '--min-area-limit', str(min_area_limit),
+  ]
+  return basic_cmd
+
+
+def run_floorplan_dse(args):
+  """Self-invoke tapac to generate multiple floorplan outputs"""
+  # remove the file opened in argparse
+  args.floorplan_output.close()
+  os.remove(args.floorplan_output.name)
+
+  args_dict = vars(args)
+  args_dict = {k:v for k, v in args_dict.items() if v is not None}
+
+  top_work_dir = args.work_dir
+  try:
+    os.makedirs(top_work_dir, exist_ok=True)
+  except:
+    _logger.error('Fail to make directory at %s. Check the --work-dir option.', top_work_dir)
+    exit(1)
+
+  _logger.info(f'')
+  _logger.info(f'-----------------------------------------------------------------------------')
+  _logger.info(f'DSE preparation: synthesize each task into RTL')
+  _logger.info(f'-----------------------------------------------------------------------------')
+  _logger.info(f'')
+
+  hls_dir = f'{os.getcwd()}/{top_work_dir}/.csynth/'
+  hls_cmd = get_cmd_to_csynth_tasks(args_dict, hls_dir, tempfile.gettempdir())
+  try:
+    subprocess.run(hls_cmd,
+                    stdout=subprocess.PIPE,
+                    check=True,
+                    universal_newlines=True)
+  except subprocess.CalledProcessError as e:
+    _logger.error('Fail to run command: %s', ' '.join(hls_cmd) )
+    exit(1)
+
+  # start exploring the pareto optimal curve
+  min_slr_limit = args.min_slr_width_limit
+  max_slr_limit = args.max_slr_width_limit
+  min_area_limit = args.min_area_limit
+  id = 1
+
+  while min_slr_limit < max_slr_limit:
+    _logger.info(f'')
+    _logger.info(f'-----------------------------------------------------------------------------')
+    _logger.info(f'Start generating design point {id} with --max-slr-width-limit {max_slr_limit}')
+    _logger.info(f'-----------------------------------------------------------------------------')
+    _logger.info(f'')
+
+    output_dir = f'{top_work_dir}/run-{id}'
+    run_dir = f'{output_dir}/run'
+
+    os.makedirs(run_dir, exist_ok=True)
+
+    # reuse the HLS and RTL synthesis results
+    try:
+      subprocess.run(['rsync', '-az', hls_dir+'/', run_dir],
+                    stdout=subprocess.PIPE,
+                    check=True,
+                    universal_newlines=True)
+    except:
+      _logger.error('Failed to run rsync from %s to %s. Check if rsync is installed', hls_dir, run_dir)
+      exit(1)
+
+    # run floorplanning
+    run_cmd = get_cmd_to_explore_floorplan(args_dict, run_dir, output_dir, max_slr_limit, min_area_limit)
+    try:
+      subprocess.run(run_cmd,
+                    stdout=subprocess.PIPE,
+                    check=True,
+                    universal_newlines=True)
+    except:
+      _logger.error('Fail to run command: %s', ' '.join(run_cmd))
+      exit(1)
+
+    # extract the floorplan metrics
+    config = json.load(open(f'{top_work_dir}/run-{id}/run/autobridge/post-floorplan-config.json', 'r'))
```

Comment:
Prefer with statement for open. Otherwise, it is undefined when the opened file will be closed.